src/Features/Core/Portable/Diagnostics/Analyzers/RemoveUnnecessaryCastDiagnosticAnalyzerBase.cs

```diff
@@ -3,76 +3,70 @@
 using System.Collections.Immutable;
 using System.Linq;
 using System.Threading;
+using Microsoft.CodeAnalysis.CodeStyle;
 using Microsoft.CodeAnalysis.Shared.Extensions;
 using Microsoft.CodeAnalysis.Text;
 
 namespace Microsoft.CodeAnalysis.Diagnostics.RemoveUnnecessaryCast
 {
-    internal abstract class RemoveUnnecessaryCastDiagnosticAnalyzerBase<TLanguageKindEnum> : DiagnosticAnalyzer, IBuiltInAnalyzer where TLanguageKindEnum : struct
+    internal abstract class RemoveUnnecessaryCastDiagnosticAnalyzerBase<
+        TLanguageKindEnum,
+        TCastExpression> : AbstractCodeStyleDiagnosticAnalyzer
+        where TLanguageKindEnum : struct
+        where TCastExpression : SyntaxNode
     {
-        private static readonly LocalizableString s_localizableTitle = new LocalizableResourceString(nameof(FeaturesResources.Remove_Unnecessary_Cast), FeaturesResources.ResourceManager, typeof(FeaturesResources));
-        private static readonly LocalizableString s_localizableMessage = new LocalizableResourceString(nameof(WorkspacesResources.Cast_is_redundant), WorkspacesResources.ResourceManager, typeof(WorkspacesResources));
+        protected RemoveUnnecessaryCastDiagnosticAnalyzerBase()
+            : base(IDEDiagnosticIds.RemoveUnnecessaryCastDiagnosticId,
+                   new LocalizableResourceString(nameof(FeaturesResources.Remove_Unnecessary_Cast), FeaturesResources.ResourceManager, typeof(FeaturesResources)),
+                   new LocalizableResourceString(nameof(WorkspacesResources.Cast_is_redundant), WorkspacesResources.ResourceManager, typeof(WorkspacesResources)))
+        {
+        }
+
+        protected abstract ImmutableArray<TLanguageKindEnum> SyntaxKindsOfInterest { get; }
+        protected abstract TextSpan GetFadeSpan(TCastExpression node);
+        protected abstract bool IsUnnecessaryCast(SemanticModel model, TCastExpression node, CancellationToken cancellationToken);
 
-        private static readonly DiagnosticDescriptor s_descriptor = new DiagnosticDescriptor(IDEDiagnosticIds.RemoveUnnecessaryCastDiagnosticId,
-            s_localizableTitle,
-            s_localizableMessage,
-            DiagnosticCategory.Style,
-            DiagnosticSeverity.Hidden,
-            isEnabledByDefault: true,
-            customTags: DiagnosticCustomTags.Unnecessary);
+        public override bool OpenFileOnly(Workspace workspace)
```

Comment:
are we sure options require for this is available in OOP? asking since it is now moved to codestyle and code style analyzer usually require workspace options.

Reply:
i'm just preserving the existing behavior here.  Do you want me to change this?  Can you refresh my memory on what this option is for and what sort of checks it should actually be doing?  I've forgotten what we've settled on as best practices here.  Thanks!

Reply:
I guess, then it probably didn't have any options before. then should be fine.